SCClassLibrary/Common/Collections/Event.sc

```diff
@@ -968,6 +968,12 @@ Event : Environment {
 							server.sendBundle(server.latency, *~bundle);
 						};
 						~bundle = nil;
+					},
+
+					composite: { |server|
+						~types.do { |type|
+							~eventTypes[type].value(server);
+						};
 					}
```

Comment:
If I understand correctly, this doesn't do what is needed to correctly call each event function. It doesn't get its `parentType` and it may mess up `~id`. I understand that this is attractive, because it is simple and efficient, but it won't be very general. 

I'd do something like:

```supercollider
composite: { |server|
	var outerEnvir = currentEnvironment;
	~types.collect { |type|
		var envir = Event.make {
			~parent = outerEnvir;
			~type = type;
		};
		envir.play;
		envir
	};
};
```

Perhaps the tests should include something like `\types, [\note, \note]`?

Reply:
That's reasonable, I'll commit an update when I can.

Reply:
Hmm... I would still like to give the user access to the events, with all of the play-time modifications.

Oh... hmm... maybe the composite event type could have a separate callback, like:

```
	composite: { |server|
		var results = ~types.collect { |type|
			Event(parent: currentEnvironment)
			.put(\type, type)
			.play;
		};
		~compositeCallback.value(results);
	}
```

```
(type: \composite, types: #[note, note], compositeCallback: { |events| events.do(_.postln) }).play;

// available to compositeCallback
( 'instrument': default, 'msgFunc': a Function, 'amp': 0.1, 'server': localhost, 
  'sustain': 0.8, 'isPlaying': true, 'freq': 261.6255653006, 'hasGate': true, 'type': note, 
  'id': [ 1000 ] )
( 'instrument': default, 'msgFunc': a Function, 'amp': 0.1, 'server': localhost, 
  'sustain': 0.8, 'isPlaying': true, 'freq': 261.6255653006, 'hasGate': true, 'type': note, 
  'id': [ 1001 ] )

// return value, untouched
-> ( 'server': localhost, 'compositeCallback': a Function, 'types': [ note, note ], 'type': composite )
```

Otherwise, we could put the collection of result events into `currentEnvironment` -- but unfortunately, that would produce an infinitely recursive data structure (because the result events have currentEnvironment as a parent). So I'm not crazy about that.

Reply:
it would also be possible to remove the parents before putting the collection of results into `currentEnvironment`?

Reply:
> it would also be possible to remove the parents before putting the collection of results into currentEnvironment?

Then the result events will contain only the intermediate values calculated each sub-event's `play` call but they won't have access to the source values. Those source values live in the parent, e.g.:

```
p = Pbind(
	\type, \composite,
	\types, #[note, midi],
	\degree, Pwhite(0, 7, inf)
);
```

Within the `composite` function, `degree` is contained in `currentEnvironment`. Then, by your suggestion, we would create a new event with `currentEnvironment` as the parent, play the new event, and then delete the parent. Then if you ask the new event for `degree`, I think you'll get nil -- because this key was in `currentEnvironment` --> new event's parent, which was removed.

I'm definitely not quite keen on having a second callback. So I'm thinking the best thing is just to copy the event, set the type and play it -- collecting those results and storing them in the event.

```
	composite: { |server|
		~resultEvents = ~types.collect { |type|
			currentEnvironment.copy.put(\type, type).play;
		};
	}
```